src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java

```diff
@@ -267,6 +289,17 @@ public MqttClient publish(String topic, Buffer payload, MqttQoS qosLevel, boolea
 
     io.netty.handler.codec.mqtt.MqttMessage publish = MqttMessageFactory.newMessage(fixedHeader, variableHeader, buf);
 
+    switch (qosLevel) {
+      case AT_LEAST_ONCE:
+        qos1outbound.put(variableHeader.messageId(), publish);
+        countInflightQueue++;
+        break;
+      case EXACTLY_ONCE:
+        qos2outbound.put(variableHeader.messageId(), publish);
+        countInflightQueue++;
+        break;
+    }
+
     this.write(publish);
 
     if (publishSentHandler != null) {
```

Comment:
But, actually, we have publishCompleteHandler which called when PUBACK or PUBCOM is received. About publishSentHandler ... I think that it would better to just remove it from params... it seems that it is useless because it can be replaced with some code after the method call.